File: LegacySource/settlement/room/law/court/CourtInstance.cs
using System;
using System.Collections.Generic;

namespace Settlement.Room.Law.Court
{
    public class CourtInstance : RoomInstance, ROOM_SERVICER
    {
        private static readonly long serialVersionUID = 1L;
        private short executions;
        private short workCurrent;
        private readonly short total;
        private readonly short[] cellsXY;
        private short cellI = 0;
        private short wI = 0;
        public readonly RoomServiceInstance service;

        protected CourtInstance(ROOM_COURT b, TmpArea area, RoomInit init) : base(b, area, init)
        {
            int spots = 0;
            foreach (COORDINATE c in body())
            {
                if (!is(c))
                    continue;
                if (CourtStation.isJudge(c))
                {
                    spots++;
                }
            }

            cellsXY = new short[spots * 2];
            total = (short)spots;
            executions = 0;
            spots = 0;
            int sers = 0;
            foreach (COORDINATE c in body())
            {
                if (is(c))
                {
                    if (CourtStation.isJudge(c))
                    {
                        cellsXY[spots++] = (short)c.x();
                        cellsXY[spots++] = (short)c.y();
                    }
                    if (Service.Init(c.x(), c.y()) != null)
                        sers++;
                }
            }
            employees().maxSet(total);
            employees().neededSet(total);
            service = new RoomServiceInstance(sers, blueprintI().data);
            foreach (COORDINATE c in body())
            {
                if (is(c))
                {
                    Service.InitInit(c.x(), c.y(), this);
                }
            }
            activate();
        }

        protected override void loadFix()
        {
            base.loadFix();
        }

        public int total()
        {
            return total;
        }

        public int executions()
        {
            return executions;
        }

        public void inc(int executions, int workCurrent)
        {
            this.executions += executions;
            this.workCurrent += workCurrent;
            if (active())
            {
                blueprintI().incPrisoners(executions, 0);
            }
        }

        protected override bool render(Renderer r, ShadowBatch shadowBatch, RenderData.RenderIterator it)
        {
            it.lit();
            return base.render(r, shadowBatch, it);
        }

        protected override void activateAction()
        {
            blueprintI().incPrisoners(executions, total);
            foreach (COORDINATE c in body())
            {
                if (is(c))
                {
                    Service s = Service.Init(c.x(), c.y());
                    if (s != null)
                    {
                        s.activate();
                    }
                }
            }
        }

        protected override void deactivateAction()
        {
            blueprintI().incPrisoners(-executions, -total);
            foreach (COORDINATE c in body())
            {
                if (is(c))
                {
                    Service s = Service.Init(c.x(), c.y());
                    if (s != null)
                    {
                        s.deactivate();
                    }
                }
            }
        }

        protected override void updateAction(double updateInterval, bool day)
        {
            if (day)
                service.updateDay();
        }

        protected override void dispose()
        {
            foreach (COORDINATE c in body())
            {
                if (!is(c))
                    continue;
                FSERVICE s = Service.Init(c.x(), c.y());
                if (s != null && s.findableReservedCanBe())
                    s.findableReserve();
            }
            service.dispose(blueprintI().data);
        }

        public override ROOM_COURT blueprintI()
        {
            return (ROOM_COURT)blueprint();
        }

        public CourtStation reserveSpot()
        {
            if (executions == total)
                throw new Exception();
            if (!active())
                throw new Exception();
            for (int i = 0; i < cellsXY.Length; i += 2)
            {
                cellI += 2;
                if (cellI >= cellsXY.Length)
                    cellI = 0;
                int tx = cellsXY[cellI];
                int ty = cellsXY[cellI + 1];
                CourtStation s = CourtStation.Init(tx, ty);
                if (s.criminalReseveredCanBe())
                {
                    s.criminalReserve();
                    return s;
                }
            }
            throw new Exception();
        }

        public CourtStation work()
        {
            if (workCurrent == 0)
                return null;
            for (int i = 0; i < cellsXY.Length; i += 2)
            {
                wI += 2;
                if (wI >= cellsXY.Length)
                    wI = 0;
                int tx = cellsXY[wI];
                int ty = cellsXY[wI + 1];
                CourtStation s = CourtStation.Init(tx, ty);
                if (s.workReservedCanBe())
                {
                    s.workReserve();
                    return s;
                }
            }
            throw new Exception();
        }

        public override RoomServiceInstance service()
        {
            return service;
        }

        public override double quality()
        {
            return ROOM_SERVICER.defQuality(this, (double)employees().employed() / employees().max());
        }
    }
}

File: settlement/room/law/court/CourtInstance.java
package settlement.room.law.court;

import settlement.misc.util.FSERVICE;
import settlement.room.main.RoomInstance;
import settlement.room.main.TmpArea;
import settlement.room.main.util.RoomInit;
import settlement.room.service.module.ROOM_SERVICER;
import settlement.room.service.module.RoomServiceInstance;
import snake2d.Renderer;
import snake2d.util.datatypes.COORDINATE;
import util.rendering.RenderData;
import util.rendering.ShadowBatch;

final class CourtInstance extends RoomInstance implements ROOM_SERVICER{

	private static final long serialVersionUID = 1L;
	private short executions;
	private short workCurrent;
	private final short total;
	private final short[] cellsXY;
	private short cellI = 0;
	private short wI = 0;
	final RoomServiceInstance service;
	
	
	protected CourtInstance(ROOM_COURT b, TmpArea area, RoomInit init) {
		super(b, area, init);
		
		int spots = 0;
		for (COORDINATE c : body()) {
			if (!is(c))
				continue;
			if (CourtStation.isJudge(c)) {
				spots ++;
			}
		}
		
		cellsXY = new short[spots*2];
		total = (short) spots;
		executions = 0;
		spots = 0;
		int sers = 0;
		for (COORDINATE c : body()) {
			if (is(c)) {
				if (CourtStation.isJudge(c)) {
					cellsXY[spots++] = (short) c.x();
					cellsXY[spots++] = (short) c.y();
				}
				if (Service.init(c.x(), c.y()) != null)
					sers++;
			}
		}
		employees().maxSet(total);
		employees().neededSet(total);
		service = new RoomServiceInstance(sers, blueprintI().data);
		for (COORDINATE c : body()) {
			if (is(c)) {
				Service.initInit(c.x(), c.y(), this);
			}
		}
		activate();
	}
	
	
	
	@Override
	protected void loadFix() {
		super.loadFix();
	}
	
	public int total() {
		return total;
	}
	
	public int executions() {
		return executions;
	}
	
	void inc(int executions, int workCurrent) {
		this.executions += executions;
		this.workCurrent += workCurrent;
		if (active()) {
			blueprintI().incPrisoners(executions, 0);
		}
	}
	
	
	@Override
	protected boolean render(Renderer r, ShadowBatch shadowBatch, RenderData.RenderIterator it) {
		it.lit();
		return super.render(r, shadowBatch, it);
	}

	@Override
	protected void activateAction() {
		blueprintI().incPrisoners(executions, total);
		for (COORDINATE c : body()) {
			if (is(c)) {
				Service s = Service.init(c.x(), c.y());
				if (s != null) {
					s.activate();
				}
			}
		}
	}

	@Override
	protected void deactivateAction() {
		blueprintI().incPrisoners(-executions, -total);
		for (COORDINATE c : body()) {
			if (is(c)) {
				Service s = Service.init(c.x(), c.y());
				if (s != null) {
					s.deactivate();
				}
			}
		}
	}

	@Override
	protected void updateAction(double updateInterval, boolean day) {
		if (day)
			service.updateDay();
	}

	
	@Override
	protected void dispose() {
		
		for (COORDINATE c : body()) {
			if (!is(c))
				continue;
			FSERVICE s = Service.init(c.x(), c.y());
			if (s != null && s.findableReservedCanBe())
				s.findableReserve();
		}
		service.dispose(blueprintI().data);
	}

	@Override
	public ROOM_COURT blueprintI() {
		return (ROOM_COURT) blueprint();
	}
	
	CourtStation reserveSpot() {
		if (executions == total)
			throw new RuntimeException();
		if (!active())
			throw new RuntimeException();
		for (int i = 0; i < cellsXY.length; i+=2) {
			cellI += 2;
			if (cellI >= cellsXY.length)
				cellI = 0;
			int tx = cellsXY[cellI];
			int ty = cellsXY[cellI+1];
			CourtStation s = CourtStation.init(tx, ty);
			if (s.criminalReseveredCanBe()) {
				s.criminalReserve();
				return s;
			}
		}
		throw new RuntimeException();
	}
	
	CourtStation work() {
		if (workCurrent == 0)
			return null;
		for (int i = 0; i < cellsXY.length; i+=2) {
			wI += 2;
			if (wI >= cellsXY.length)
				wI = 0;
			int tx = cellsXY[wI];
			int ty = cellsXY[wI+1];
			CourtStation s = CourtStation.init(tx, ty);
			if (s.workReservedCanBe()) {
				s.workReserve();
				return s;
			}
		}
		throw new RuntimeException();
	}



	@Override
	public RoomServiceInstance service() {
		return service;
	}



	@Override
	public double quality() {
		return ROOM_SERVICER.defQuality(this, (double)employees().employed()/employees().max());
	}
	
	

}
